File: myfirstmod/src/main/java/com/example/dartbridge/proxy/DartBlockProxy.java
package com.example.dartbridge.proxy;

import com.example.dartbridge.DartBridge;
import net.minecraft.core.BlockPos;
import net.minecraft.world.InteractionResult;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.block.Block;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.world.phys.BlockHitResult;

/**
 * A Block proxy that delegates all behavior to Dart.
 *
 * Each instance of this class represents a single Dart-defined block type.
 * The dartHandlerId links to the Dart-side CustomBlock instance.
 */
public class DartBlockProxy extends Block {
    private final long dartHandlerId;

    public DartBlockProxy(Properties settings, long dartHandlerId) {
        super(settings);
        this.dartHandlerId = dartHandlerId;
    }

    public long getDartHandlerId() {
        return dartHandlerId;
    }

    @Override
    public BlockState playerWillDestroy(Level level, BlockPos pos, BlockState state, Player player) {
        // Delegate to Dart
        if (DartBridge.isInitialized()) {
            DartBridge.onProxyBlockBreak(
                dartHandlerId,
                level.hashCode(),
                pos.getX(),
                pos.getY(),
                pos.getZ(),
                player.getId()
            );
        }
        return super.playerWillDestroy(level, pos, state, player);
    }

    @Override
    protected InteractionResult useWithoutItem(BlockState state, Level level, BlockPos pos,
                                                Player player, BlockHitResult hit) {
        if (!DartBridge.isInitialized()) {
            return InteractionResult.PASS;
        }

        int result = DartBridge.onProxyBlockUse(
            dartHandlerId,
            level.hashCode(),
            pos.getX(),
            pos.getY(),
            pos.getZ(),
            player.getId(),
            0  // hand ordinal - simplified for now
        );

        // Map result ordinal to InteractionResult
        // In 1.21+, InteractionResult is simplified
        return switch (result) {
            case 0 -> InteractionResult.SUCCESS;        // success, arm swings
            case 1, 2 -> InteractionResult.CONSUME;     // consume variants
            case 3 -> InteractionResult.PASS;           // no interaction
            case 4 -> InteractionResult.FAIL;           // interaction failed
            default -> InteractionResult.PASS;
        };
    }
}

File: dart_mc_bridge/fabric_bridge/src/main/kotlin/com/example/dartbridge/DartBlockProxy.kt
package com.example.dartbridge

import net.minecraft.block.Block
import net.minecraft.block.BlockState
import net.minecraft.block.Blocks
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.util.ActionResult
import net.minecraft.util.Hand
import net.minecraft.util.hit.BlockHitResult
import net.minecraft.util.math.BlockPos
import net.minecraft.world.World

/**
 * A proxy block that delegates its behavior to Dart code.
 *
 * This block can be used to create custom blocks where the logic
 * is implemented in Dart rather than Kotlin/Java.
 *
 * Usage:
 * 1. Register this block with a unique identifier
 * 2. In Dart, register handlers for this block's events
 * 3. When players interact with the block, Dart handlers are called
 */
class DartBlockProxy(settings: Settings) : Block(settings) {

    companion object {
        /**
         * Create a DartBlockProxy with default settings.
         */
        fun create(): DartBlockProxy {
            return DartBlockProxy(
                Settings.copy(Blocks.STONE)
                    .strength(2.0f, 6.0f)
            )
        }

        /**
         * Create a DartBlockProxy copying settings from another block.
         */
        fun copyOf(block: Block): DartBlockProxy {
            return DartBlockProxy(Settings.copy(block))
        }
    }

    // TODO: Add block ID for Dart to identify which block is being interacted with
    // This could be used to have multiple different Dart-controlled blocks

    @Deprecated("Deprecated in Java")
    override fun onUse(
        state: BlockState,
        world: World,
        pos: BlockPos,
        player: PlayerEntity,
        hand: Hand,
        hit: BlockHitResult
    ): ActionResult {
        if (world.isClient) {
            return ActionResult.SUCCESS
        }

        if (!DartBridge.isInitialized()) {
            return ActionResult.PASS
        }

        val handValue = if (hand == Hand.MAIN_HAND) 0 else 1
        val result = DartBridge.onBlockInteract(
            pos.x,
            pos.y,
            pos.z,
            player.id.toLong(),
            handValue
        )

        return if (result != 0) {
            ActionResult.SUCCESS
        } else {
            ActionResult.FAIL
        }
    }

    override fun onBreak(world: World, pos: BlockPos, state: BlockState, player: PlayerEntity): BlockState {
        if (!world.isClient && DartBridge.isInitialized()) {
            val result = DartBridge.onBlockBreak(
                pos.x,
                pos.y,
                pos.z,
                player.id.toLong()
            )

            if (result == 0) {
                // Dart cancelled the break - we can't actually prevent it here
                // since onBreak is called after the break is confirmed,
                // but we could add visual/audio feedback
                println("[DartBlockProxy] Block break was marked as cancelled by Dart")
            }
        }

        return super.onBreak(world, pos, state, player)
    }
}
